pkg/helm/charts.go

```diff
@@ -302,6 +302,43 @@ func (c *Charts) AddRepos(repos ...Repo) error {
 	return nil
 }
 
+// VersionCheck checks each of the charts in the requires section and returns if there are any
+// newer versions of the chart available on the repository.
+func (c *Charts) VersionCheck(repoConfigPath string) (ChartSearchVersions, error) {
+	chartVersions := ChartSearchVersions{}
+
+	if repoConfigPath != "" {
+		repoConfig, err := LoadHelmRepoConfig(repoConfigPath)
+		if err != nil {
+			return nil, err
+		}
+		c.Manifest.Repositories = repoConfig.Repositories
+	}
+
+	for _, r := range c.Manifest.Requires {
+		searchVersions, err := c.Helm.SearchRepo(r.Chart, r.Version, Opts{Repositories: c.Manifest.Repositories})
+		if err != nil {
+			return nil, err
+		}
+
+		// Since we can store multiple versions of the same chart in the chartfile we dedup here.
+		for _, sv := range searchVersions {
+			exists := false
+			for _, cv := range chartVersions {
+				if sv == cv {
+					exists = true
+					break
+				}
+			}
+			if !exists {
+				chartVersions = append(chartVersions, sv)
+			}
+		}
```

Comment:
I'm wondering if keeping the same order as the requires would be easier to work with

The main goal is to enable our internal workflow:

- As a first step, we probably want to abort the process if there's more than one required version for the same chart already. This may mean that an upgrade is already in progress OR the environment is manually overridden. By having one item per "require", we have this information in the version check command
- By having one version check array item per requirement, we can more safely modify the chartfile. We know that the version check item index will match what's in the chartfile

So I'd propose the following format:

```
[
  {
     name
     description
     directory (etc, all items of the require)
     current_version: { chart_version, app_version }
     latest_version: { chart_version, app_version }
     (if possible) latest_matching_major_version: { chart_version, app_version }
     (if possible) latest_matching_minor_version: { chart_version, app_version }
   }
]
```
with one item per "require"

I think this gives us more options to work with afterwards. What do you think?

Reply:
I definitely like the format to provide more information about each upgrade.

As for exiting the process early if there are multiple versions of the same `require`, I'm not entirely sure. We have a few `chartfile.yaml` files that contain multiple charts, and exiting early would mean that we could only be upgrading a single chart at a time. Additionally, unless we had a random choice or something, it could only ever upgrade the first chart in the list if they are releasing faster than we are running our cron. With that in mind, that's why I went with the "just return the info and let the level above figure out what should be upgraded" rather than baking that decision into tanka.

With that in mind, here's a scenario:
```yaml
# chartfile.yaml
...
requires:
  - chart: registry/chart1
    version: 1.0.0
  - chart: registry/chart1
    version: 1.0.1
  - chart: registry/chart2
    version: 3.0.0
```

If `chart1` latest is `1.0.1`, and `chart2` latest is `3.0.1`, the return I think would be the most amount of information would be
```json
{
  "registry/chart1@1.0.0": {"name": "registry/chart1", "latest_version": "1.0.1", ...},
  "registry/chart1@1.0.1": {},
  "registry/chart2@3.0.0": {"name": "registry/chart2", "latest_version": "3.0.1", ...},
}
```

Which would allow the higher-level tool for each require to do a quick `if {name}@{latest_version} in requires` before deciding to do an upgrade.

We would lose out on the list-ordering here by returning the object but I think the output of this will either be used by humans who could add to the requires how they wish, or used by robots where humans no longer have to really manage the `chartfile` and don't care to maintain the order of the list.

I would have all that info you had in yours in those returned objects, I just cut them short for clarity here.

What are your thoughts about that?